apps/frontend/src/lib/storage/localStorage/user/connection/message/index.ts

```diff
@@ -0,0 +1,151 @@
+    // Ensure messages are in order and update lastMessageTimestamp
+    if (lastMessageTimestamp && message.createdAt < lastMessageTimestamp) {
+      throw new Error("Messages are not in order");
+    }
+    lastMessageTimestamp = message.createdAt;
+
+    try {
+      const { serializedData, senderSignaturePublicKey } =
+        await decryptReceivedMessage({
+          messageData: message,
+          encryptionPrivateKey: user.encryptionPrivateKey,
+        });
+
+      // Only handle tap back messages for now
+      // TODO: Handle other message types
+      const tapBackMessage = parseSerializedTapBackMessage(serializedData);
+      if (Object.keys(connectionTapBacks).includes(senderSignaturePublicKey)) {
+        connectionTapBacks[senderSignaturePublicKey] = [
+          ...connectionTapBacks[senderSignaturePublicKey],
+          tapBackMessage,
+        ];
+      } else {
+        connectionTapBacks[senderSignaturePublicKey] = [tapBackMessage];
+      }
+    } catch (error) {
+      console.error("Error processing message:", error);
+    }
+  }
+
+  if (!lastMessageTimestamp) {
+    throw new Error("No messages found");
+  }
+
+  const newBackupEntries: CreateBackupData[] = [];
+
+  Object.entries(connectionTapBacks).forEach(
+    ([senderSignaturePublicKey, tapBackMessages]) => {
+      if (tapBackMessages.length === 0) {
+        return;
+      }
+
+      const connectionUsername = tapBackMessages[0].user.username;
+      let updatedConnection: Connection = user.connections[connectionUsername];
+
+      // Create new connection if it doesn't exist
+      if (!updatedConnection) {
+        const firstTapBackMessage = tapBackMessages[0];
+        const newConnection: Connection = {
+          user: firstTapBackMessage.user,
+          taps: [],
+          sentMessages: [],
+        };
+        updatedConnection = newConnection;
+      }
+
+      // Update connection with new tap back messages
+      for (const tapBackMessage of tapBackMessages) {
+        // Check claimed signature public key
+        if (
+          senderSignaturePublicKey !== tapBackMessage.user.signaturePublicKey
+        ) {
+          throw new Error("Signature public key does not match");
+        }
+
+        // Do not allow a connection to change username or keys
+        if (
+          updatedConnection.user.username !== tapBackMessage.user.username ||
+          updatedConnection.user.signaturePublicKey !==
+            tapBackMessage.user.signaturePublicKey ||
+          updatedConnection.user.encryptionPublicKey !==
+            tapBackMessage.user.encryptionPublicKey ||
+          updatedConnection.user.psiPublicKeyLink !==
+            tapBackMessage.user.psiPublicKeyLink
+        ) {
+          throw new Error("Connection cannot change username or keys");
+        }
+
+        updatedConnection = {
+          ...updatedConnection,
+          user: {
+            ...updatedConnection.user,
+            ...tapBackMessage.user,
```

Comment:
I'm not familiar with this syntax -- I'm assuming user is first set to values from updatedConnection and then tapBackMessage overrides non-matching values?

Reply:
yup